core/session.ts

```diff
@@ -445,33 +393,134 @@ export class Session {
     });
   }
 
-  public checkTargetIsUnused(target: dataform.ITarget) {
-    const duplicateActions = this.findActions({ schema: target.schema, name: target.name });
-    if (duplicateActions && duplicateActions.length > 0) {
-      this.compileError(
-        new Error(
-          `Duplicate action name detected. Names within a schema must be unique across tables, assertions, and operations: "${target.schema}.${target.name}"`
-        )
-      );
-    }
+  public setNameAndTarget(action: IActionProto, name: string, overrideSchema?: string) {
+    action.target = this.target(name, overrideSchema || this.config.defaultSchema);
+    action.name = `${action.target.schema}.${action.target.name}`;
   }
 
-  public getSuffixWithUnderscore() {
+  private getSuffixWithUnderscore() {
     return !!this.config.schemaSuffix ? `_${this.config.schemaSuffix}` : "";
   }
 
-  public setNameAndTarget(action: IActionProto, name: string, overrideSchema?: string) {
-    const newTarget = overrideSchema ? this.target(name, overrideSchema) : this.target(name);
-    this.checkTargetIsUnused(newTarget);
-    action.target = newTarget;
-    action.name = `${action.target.schema}.${action.target.name}`;
+  private compileGraphChunk<T>(actions: Array<{ proto: IActionProto; compile(): T }>): T[] {
+    const compiledChunks: T[] = [];
+
+    actions.forEach(action => {
+      try {
+        const compiledChunk = action.compile();
+        compiledChunks.push(compiledChunk);
+      } catch (e) {
+        this.compileError(e, action.proto.fileName);
+      }
+    });
+
+    return compiledChunks;
+  }
+
+  private target(name: string, schema: string): dataform.ITarget {
+    const adapter = this.adapter();
+    return dataform.Target.create({
+      name: adapter.normalizeIdentifier(name),
+      schema: adapter.normalizeIdentifier(schema)
+    });
+  }
+
+  private fullyQualifyDependencies(actions: IActionProto[]) {
+    const allActionsByName = keyByName(actions);
+    actions.forEach(action => {
+      action.dependencies = (action.dependencies || []).map(dependencyName => {
+        if (!!allActionsByName[dependencyName]) {
+          // Dependency is already fully-qualified.
+          return dependencyName;
+        }
+        const possibleDeps = this.findActions(dependencyName);
+        if (possibleDeps.length === 1) {
+          return possibleDeps[0].proto.name;
+        }
+        if (possibleDeps.length >= 1) {
+          this.compileError(new Error(utils.ambiguousActionNameMsg(dependencyName, possibleDeps)));
+        } else {
+          this.compileError(
+            new Error(
+              `Missing dependency detected: Action "${action.name}" depends on "${dependencyName}" which does not exist.`
+            ),
+            action.fileName
+          );
+        }
+        return dependencyName;
+      });
+    });
+  }
+
+  private appendSchemaSuffix(actions: IActionProto[]) {
```

Comment:
For my sanity, what does this function do? It doesn't return anything and every variable it changes is inside it's scope, no?

Reply:
It mutates the state of the protos which get passed into it; those protos are still in scope in the caller function.

I'm not a fan of how stateful and non-idempotent this makes `Session.compile()` (it's fairly horrible), but this doesn't make anything worse - compilation was unfortunately previously not idempotent.

FWIW making it stateless is on my wishlist for the open-source repo: https://docs.google.com/document/d/10ewZ0xhUuTVs6SHEycqTfDOyj-5UYPVfYzlWph1TpA0/edit#